### ISSUE TITLE
DeepMark

### ISSUE THREAD

**Issue by soumith** [soumith/convnet-benchmarks#101]
Hi all,

The reason I've been slow on convnet-benchmarks these days is because i've been working on the side on DeepMark.

I initially wrote *convnet-benchmarks* to increase competition among frameworks so that we can work towards faster ConvNets, and they served their purpose well. After the release of convnet-benchmarks, multiple frameworks pulled up their socks to speedup convnets, with a deep sense of prestige for being on top of these benchmarks. In these two years, we as a community accelerated GPU ConvNets across all frameworks between 4x to 10x, efficiently implementing tricks such as FFT, Winograd, and powered by faster hardware. Alex Khrizevsky, Yangqing Jia, Scott Gray, Nicolas Vasilache, Sander Dieleman, Michael Mathieu, Julien Denmouth and many other human compilers helped make this a reality -- looking at the diversity in terms of where each of us work(ed) professionally shows that this kind of acceleration was truly a community effort with a ton of openness, something that is plain awesome! :)
I've also enjoyed reading the deeply technical discussions that take place on convnet-benchmarks (my favorites: https://github.com/soumith/convnet-benchmarks/issues/93 , https://github.com/soumith/convnet-benchmarks/issues/66 , https://github.com/soumith/convnet-benchmarks/issues/59 in recent times ).

Moving on,  *convnet-benchmarks* do not accurately capture everything we think of when we say *deep learning*. We don't have Recurrent Nets, we don't have video use-cases, speech, NLP etc. There is a need for such comprehensive benchmarks, especially as the space is getting ready for dedicated hardware chips, multi-GPU and multi-machine frameworks and more complex use-cases.

I've sat down with a few of you at NIPS and GTC to discuss and freeze the initial round of benchmarks for what I am calling **DeepMark**. My initial plan was to work on the initial set of benchmark scripts by myself and cover the most popular frameworks, and then let the direction and maintenance of the benchmarks be community-driven. But the breadth of this effort has been overwhelming to say the least. After careful thought, I've decided that I'll just ask everyone to pitch in for their part of the benchmarks with making scripts etc., especially as many of you were very receptive to the idea offline.

Here are the initial set of use-cases we want to cover:

### Images
- InceptionV3-batchnorm (http://arxiv.org/abs/1512.00567 , https://github.com/Moodstocks/inception-v3.torch)
- Alexnet-OWT
- VGG
- ResNet-50 ( http://arxiv.org/abs/1512.03385 , https://github.com/facebook/fb.resnet.torch )

### Video
- C3D - A vgg-style 3D net ( http://vlg.cs.dartmouth.edu/c3d/ )

### Audio
- DeepSpeech2 - Convnet + RNN + FC ( http://arxiv.org/abs/1512.02595 )
- MSR's 5 layer FC net ( https://github.com/Alexey-Kamenev/Benchmarks )

### Text
- Small RNN LSTM ( https://github.com/karpathy/char-rnn/blob/master/train.lua#L38-L48 )
- Large RNN LSTM ( BIG-LSTM in http://arxiv.org/abs/1602.02410 )

### Platform
- Initially multi-GPU with (1 to 4 titan-X cards)
- However, multi-machine, custom hardware, other GPU cards such as AMD, CPUs etc. can and should be accommodated, we will work this out after the initial push.

## Metrics
- Round-trip time for 1 epoch of training (will define an epoch size separately for each network)
- Maximum batch-size that fits (to show and focus on the extra memory consumption that the framework uses)

## Frameworks
Everyone who wants to join-in, but I thought an initial set that is important to cover would be:
- Caffe
- Chainer
- MXNet
- Neon
- Theano
- TensorFlow
- Torch

## Scripts format
- Emit JSON output (so that the README -- or jekyll website can be auto-generated, similar to http://autumnai.com/deep-learning-benchmarks )

## Guarantees
- I will personally to the best of my abilities make sure that the benchmarking is fair and unbiased. The hope is that the community at large will watch these and point-out / fix mistakes.

## Governance
- The benchmarks will be placed at https://github.com/DeepMark/deepmark and other key community members / organizations who want ownership will be welcome to join in proposing new benchmarks that get relevant as the field progresses.

My hope is that these new set of benchmarks will not only increase competition but will also be beneficial in other ways to the community, serving as common examples to get started, etc.

Let me know what you think :)
Soumith

cc: @hughperkins @f0k @scott-gray @rajatmonga @vrv @bennane @nouiz @Yangqing @tqchen @unnonouno

**Comment by soumith** [comment]
Oh lastly, a good timeline for this would be to get an initial round of benchmarks by **June 15th** (since I only gave some of you a heads-up right now)

**Comment by daviddao** [comment]
So awesome and useful. What are the data sets one should benchmark on? ImageNet, CIFAR10? It would also be nice to compare the accuracy of current implementations for each framework (although that would probably be a lot of work).

**Comment by Smerity** [comment]
For text, I'd hope to expand beyond just RNN character generation. It doesn't capture many of the complexities of other models, such as variable sequence lengths or bidirectional RNNs.

The [Attention Sum Reader](https://www.semanticscholar.org/paper/Text-Understanding-with-the-Attention-Sum-Reader-Kadlec-Schmid/54fb788650fdbaeb5ab1401800f5d2e1f5c9b0d0) is a simple architecture (bidirectional GRU + dot product) that currently has SotA and could allow for optimizing sequences of different lengths, a major issue in RNNs. The model also has four different dataset sizes, small (Children's Book Test), medium (CNN), and large (Daily Mail), which are publicly available.

**Comment by vrv** [comment]
This is great, thanks for organizing this!  One thing I've also been thinking about like @daviddao is how to validate that the models are actually computing the same thing -- I've seen some benchmarks elsewhere that have raised personal doubts that the frameworks are computing the same function.  As part of the benchmark framework, maybe the API could include a way to validate that given specified initialization and input, the outputs (forward and backward) are approximately equal.  Open to thoughts :).  Cheers!

**Comment by craffel** [comment]
Nice!  I am very excited for this.

I have https://github.com/craffel/lstm_benchmarks, which is an out-of-date benchmark of Theano vs. rnnlib vs. currennt (which, at the time I wrote the benchmarks, were essentially the only options for LSTM).  The task was CHIME noisy speech recognition, which has pretty limited adoption so I would not strongly advocate for it being added as a task.  And I assume that rnnlib and current shouldn't be included in these benchmarks are they are RNN-only, right?

I'll be happy to contribute to some of the Theano RNN benchmarks once it becomes appropriate to do so.

**Comment by hughperkins** [comment]
Caveat: per Paul Graham, better to go deep, do something very well, than kind of blur one's 'focus' over many things.  I worry gently that if too many benchmarks then:
- each benchmark less well maintained
- more confusing to read

**Comment by f0k** [comment]
:+1:

> What are the data sets one should benchmark on? ImageNet, CIFAR10?

Training on something like ImageNet would move away the focus from pure computation to fast dataset iteration -- this would be interesting as well, but should probably become a separate benchmark since not all frameworks actually provide any tools for this. The other extreme would be training on random dummy data (like sampled from a Gaussian), but this makes sense only if we can guarantee the running time does not depend on the input data. So probably we should have some realistic set of inputs for each task, just large enough to fill two batches or so?

> As part of the benchmark framework, maybe the API could include a way to validate that given specified initialization and input, the outputs (forward and backward) are approximately equal.

This seems useful. It requires initial model parameters to be dumped in some format and loaded into each framework, but it would help to ensure that all implementations are the same.

**Comment by hughperkins** [comment]
> This seems useful. It requires initial model parameters to be dumped in some format and loaded into each framework, but it would help to ensure that all implementations are the same.

In the strongest case, weight initialization could be defined precisely as:
- a precise order of initialization, eg by layer, then by infeatureplane, then by outfeatureplane, then by height, then by width (for example)
- a precise random function to use (eg mt19937)
- the exact seed to use

**Comment by f0k** [comment]
> In the strongest case, weight initialization could be defined precisely as [...]

I guess getting the same stream of pseudo-random values in all different frameworks is more difficult than importing a set of tensors into all different frameworks. We wouldn't want to exclude candidates from being benchmarked because they fail to implement the same RNG.

**Comment by hughperkins** [comment]
> I guess getting the same stream of pseudo-random values in all different frameworks is more difficult than importing a set of tensors into all different frameworks. We wouldn't want to exclude candidates from being benchmarked because they fail to implement the same RNG.

Having gone through the exact same process, to compare DeepCL with convnetjs, I found it significantly easier to make use the exact same weight generator in javascript for convnetjs, than to load weights from a file https://github.com/hughperkins/DeepCL/blob/master/prototyping/convnetjs-reference/testconvnet2.js#L143-L201 .  It was a long time ago, so I dont remember why.  I do remember I initially tried writing weights to a file though, and I couldnt get it to work as easily as syncing weight generators, for some reason.

**Comment by f0k** [comment]
> I found it significantly easier to make convnetjs use the exact same weight generator as DeepCL, than to load weights from a file

If that's the case, one could of course create the initial weights in a reproducible way *and* save them to files, so implementers for the different benchmarked frameworks can choose whatever is easiest.
(Note that loading from files has the additional benefit of documenting how to load foreign model parameters into each framework.)
Umm... are we supposed to discuss such details here or over at the deepmark repository?

**Comment by soumith** [comment]
@daviddao @vrv @hughperkins @f0k for V1, I thought we should just go with synthetic data. It's very hard to setup to-convergence benchmarks, as there are very fine details wrt convergence being guaranteed, for example: some of the models (like googlenetv3) have taken a year to reproduce outside of the paper.

@Smerity In terms of evaluating perf, we can add a bidirectional RNN too. In fact, DeepSpeech2 has bidirectional-RNNs, so that should be sufficient?

@vrv definitely a great idea, but very very hard and takes a ton of resources. I feel like atleast for V1, we should just go with code review + synthetic data.

@craffel Awesome! At the moment I dont have a Point of Contact for Theano, maybe a combination of you, @f0k and @benanne could work (especially if they're implemented in Lasagne).

**Comment by craffel** [comment]
> (especially if they're implemented in Lasagne).

That would be nice :) though I am personally interested in which of the Theano-based libraries manage to eek out the most performance, since their implementations are nonidentical.

**Comment by Yangqing** [comment]
Thanks @soumith for organizing this effort! I think this would definitely help us advance the field to the next level.

I am also very interested in benchmarking not only the training pipeline, but a more wide range of evaluation criteria. The reason is as follows: if I may make a bold claim, I believe that all frameworks will again very quickly converge to the same performance, because there is no fundamental difference between them. What we saw at convnet-benchmark is that almost everyone is using the same underlying library, and we are effectively benchmarking framework overheads, something that is good to know of course, but seems to be overwhelmed by other factors, such as ease to use etc.

Given the wide attention of benchmarks, I think it would be great if we can draw attention to some of the more practical issues, such as small batch sizes in deployment time - several frameworks (including some non-open-source production systems I've worked on) have historically ignored this, and I think it is worthwhile to invite people to invest more on this direction.

I have not got a perfece idea on this, of course. One thing we can do is to simply benchmark different batch sizes, but a more complex, and potentially useful, way is probably to set up a harness that can simulate requests generated from a Poisson distribution and comes with latency requirements, and see whether frameworks can address that in an optimal fashion - this might be too application specific, though. Just my 2 cents.

(Also adding @ajtulloch to the conversation.)

**Comment by moloned** [comment]
How about per-pixel scene labelling and optical flow?

Regards

-David

**Comment by tqchen** [comment]
I agree that evaluate more aspects. Some of them is already covered in this proposal, for example
- Memory consumption
- Parallelization and Scheduling overhead

One of the most important factor, tradeoff between ease of use and optimization, was unfortunately not easy to benchmark as each people have their own taste.

What @Yangqing suggested is more on measuring perf for production and serving pipeline, which could be a whole area of new directions. As this benchmark was primarily on training. One alternative could be making a deep-serving benchmark on DeepMark organization that dedicate to this topic.

**Comment by moloned** [comment]
It would be interesting to log the power dissipation in each testcase, as
well as fps, memory BW, FLOPS etc.

**Comment by hughperkins** [comment]
> It would be interesting to log the power dissipation in each testcase

I like this idea.  A Titan draws 250watts peak (I think?).  24 hours a day for a year, 250watts is about ~600usd, which is in the same order of magnitude as the purchase price.

And power dissipation is going to become the main bottleneck plausibly in years to come. ("And over here we have our farm of 1000 Titan 2026s, and over there is the 20MW pebble bed we are using to power them" :-) )

**Comment by forresti** [comment]
@Yangqing 
>"if I may make a bold claim, I believe that all frameworks will again very quickly converge to the same performance, because there is no fundamental difference between them."

Agreed. Soumith's current benchmarks are useful, but they mainly evaluate "who can make the thinnest wrapper around cuDNN, Neon, or similar?"

It would be useful to benchmark implementations of emerging algorithms for which tuned libraries may not yet exist -- certain versions of LSTMs and RNNs, for instance.

**Comment by soumith** [comment]
@forresti yea, for historical context it used to be not like that, but it is like that now. I think for LSTMs and RNNs, a lot of perf is still up for grabs.

**Comment by hughperkins** [comment]
> Agreed. Soumith's current benchmarks are useful, but they mainly evaluate "who can make the thinnest wrapper around cuDNN, Neon, or similar?"

To be fair, cudnn, neon are competing with each other.  The opencl implementations mostly copy the caffe cuda implementation of im2col as far as I know :-D  but have different performance from cudnn.  There is also 16-bit vs 32-bit.

**Comment by forresti** [comment]
@hughperkins Good point. I didn't mean to imply that there isn't a diverse array of low-level computational libraries for DNNs.

To tune up my comment a bit: "When doing speed/efficiency benchmarks, it's hard to avoid conflating low-level computational libraries (cuDNN, Neon, various OpenCL efforts, ...) and higher-level frameworks (Caffe, Torch, Tensorflow, Theano, ...)."

**Comment by scott-gray** [comment]
I would say that convolution is far from a solved problem.  I still have a long list of optimizations I want to make.  The biggest area to explore is how to best leverage lower precision without sacrificing accuracy.  The obvious target there would be xnor nets but maybe a bit more precision is best for the highest levels of accuracy.  The 4x int8 performance that Pascal will soon have (unfortunately not in P100 though) is a tantalizing format to target.  And also obviously the native fp16 support.

Another area is better efficiency at smaller batch sizes.  I have some brand new work there that I'd like to show off.  This is important for both inference and scaling to many nodes.

Power comparisons are useful but only when looking at implementations that have the same computational throughput.  Or just use some kind of flops/watt metric.  With my newer kernels I'm getting very good at squeezing the most out of cache utilization and hence I'm hitting and maintaining higher boost clocks (while using smaller and more versatile tiles).

As for the frameworks, the big area to focus on is graph based optimizations.  Maximizing data locality (compounding), memory allocation vs compute trade-offs, auto-parallelizing independent work across streams and gpus, and lots of other creative things computational graphs greatly simplify.

As for synthetic vs real data and parameters.. In fp32 I think only the distribution matters for performance comparisons.  But in lower precision like fp16 it's very easy to saturate or underflow with synthetic data which leads to far higher performance than is warranted.  At the very least you want to account for the fan-in when setting weight magnitudes (Kaiming, Xavier, etc).  Batch norm helps a lot here too.  Basically you should be able to prove that you can train with the params you benchmark with.

At the end of the day we care about speed and usability.   I think these benchmarks should make both pretty clear.  For usability you'll be able to inspect the script to see who has the cleanest syntax and solves the most problems for you without extra steps.

**Comment by forresti** [comment]
@scott-gray That all sounds great. :)

**Comment by hughperkins** [comment]
> just use some kind of flops/watt metric

Well, the ideal would be joules per batch.  But I think this will be tricky to measure.  Might need some specialized hardware device, that sits on the power bus?

**Comment by scott-gray** [comment]
Maybe it wouldn't be quite so tricky.  You'd just need to collect some running average of the on chip power stats during the execution of the epoc.  Something like this would give you realtime stats:

```nvidia-smi -i 1 --loop-ms=333 --format=csv,noheader --query-gpu=power.draw,clocks.gr,temperature.gpu,fan.speed,clocks_throttle_reasons.sw_power_cap,clocks_throttle_reasons.hw_slowdown```

Or even better tie your benchmark script directly into NVML queries: 
https://developer.nvidia.com/nvidia-management-library-nvml

But I guess you'd want to be running these queries continuously so maybe as a separate process would be better.  You'd just need to synchronize the collection with the execution of the network.  Just a small bit of shell scripting should achieve this.

**Comment by hughperkins** [comment]
> Or even better tie your benchmark script directly into NVML queries:
https://developer.nvidia.com/nvidia-management-library-nvml

Interesting.  Seems it's just a c-interface, so accessible using ffi etc.

```
nvmlDeviceGetPowerUsage(nvmlDevice_t device, unsigned int *power);
```

**Comment by scott-gray** [comment]
And python bindings can be found here:
https://pypi.python.org/pypi/nvidia-ml-py

**Comment by scott-gray** [comment]
But, it's worth pointing out that the boost clock is already tightly coupled with these real-time power and temperature measurements so the overall timings should be reflective of this.  So perhaps it's not worth the effort.

**Comment by naibaf7** [comment]
@soumith 
Great! I'll make sure to finish my OpenCL cuDNN replacement before initial benchmarking.
Can I ask to include also AMD and Intel GPU (+CPU) benchmarks (where applicable and possible)?

**Comment by soumith** [comment]
@naibaf7 some of the benchmarks wont be able to run on the AMD cards which right now seem to be limited to 4GB variants. I have a Fury Nano that I've started playing with. Are Intel GPU benchmarks even relevant? I can run Intel GPU benchmarks on like a macbook pro or something. Intel CPU, def can run.

**Comment by naibaf7** [comment]
@soumith 
A single buffer can't be bigger than 1/4th of the GPUs memory on current AMD GPUs. I think the Fury Nano is not the right candidate to benchmark with.

I have a set of W9100 cards with 16 GB memory that would be more suitable to benchmark AMD.

@soumith please see my updated post above for additional information :)
Regards

Fabian

**Comment by naibaf7** [comment]
My idea to include a U-Net like architecture also ties in with @moloned 's suggestion.

**Comment by hughperkins** [comment]
@naibaf7 

W9100s are 2500usd each.  You'd need performance to be ~2.5 times faster than Titan X for the price/performance ratio to be attractive.  R9 Nanos are 500usd a pop, so you could get two for the price of one Titan X, giving you 8GB total memory, and then use alexnet-style parallelism over these?

**Comment by naibaf7** [comment]
@hughperkins 
If we limit the memory usage to 8 GB the W9100 should perform similarly to R9 390X, which have even higher clock speeds than a W9100 and are clocking in at only 220usd each, roughly 10x less than a W9100.

**Comment by hughperkins** [comment]
Yes. r9-390x are nice :-)

**Comment by naibaf7** [comment]
@soumith 
Oh right. My fault on the price conversion. Corrected it above.

**Comment by hughperkins** [comment]
> re: power consumption

Ah, hmmm, thats a good point. looks like one r9-390x uses about the same power as a titan x http://www.amazon.com/s/ref=nb_sb_noss_2?url=search-alias%3Daps&field-keywords=titan+X&rh=i%3Aaps%2Ck%3Atitan+X  Anyway, seems like a nice card to benchmark on, as long as the benchmarks are tweaked to fit in memory (most of them will run in 8GB I think, with slightly smaller batch sizes?)

**Comment by nervanasys** [comment]
Do you have a reference to the paper for MSR's 5 layer FC net?

**Comment by ajtulloch** [comment]
AFAIK it's just a model for benchmarking, probably derived from some of their fully-connected speech recognition nets. It's similar to the net MSR used in the 1-bit SGD paper for speech recognition (http://research.microsoft.com/pubs/230137/IS140694.PDF), but not identical.

**Comment by hughperkins** [comment]
Observations:
- I reckon it'd be cool if the benchmarks are automated, so we can submit a github repo, and a branch name, and push changes to that, and the results will either be generated automatically, or eg once a week.
- Personally, I doubt I'm going to be able to support so many benchmarks (at least initially), so I think it'd be nice if eg a framework could submit initially only for eg image recognition benchmarks, and then could elect at some later date to submit to one or more additional domains.

**Comment by pranv** [comment]
> but from my own experience with the LSTM benchmark it can be very difficult - you have to make sure literally every hyperparameter is identical, and you effectively can't use any RNGs. 

@craffel, couldn't get what do you meant by this. If we are not worried about convergence and look at only the speed, I don't see a real issue with hyperparameters as long as all nets are implemented based on the same pseudocode. Also, I think over time many frameworks will shift to using cuDNN RNN primitives.

I hope there can be some way of comparing the extent of support for recurrence offered by the framework - like presence of `scan(...)`, handling variable length sequences etc

About comparing usability: I  think its better not to compare frameworks wrt usability as there is no simple way to quantify the variations among frameworks. Syntax could look clean because of a lot of in-built APIs for standard things **now** but for "research", these may not help much. Mental overhead while writing these scripts also may have to be considered. So, it is not easy to capture and convey all of these aspects in a meaningful and an unbiased way. In the list of frameworks above, all but Caffe have a numpy-like API for manipulating Tensors in the most common ways and a mention of it somewhere should suffice.

**Comment by craffel** [comment]
> One thing I've also been thinking about like @daviddao is how to validate that the models are actually computing the same thing -- I've seen some benchmarks elsewhere that have raised personal doubts that the frameworks are computing the same function.

**Comment by pranv** [comment]
@craffel, I thought you were saying something specific about LSTM, sorry, my bad.

**Comment by andravin** [comment]
@soumith thank you for continuing to lead this valuable benchmark project, which benefits the entire machine learning community.

+1 for Intel GPUs. They are potentially very relevant for inference. In a few weeks you will be able to buy an Intel Skull Canyon NUC system for $999 that has 2.3 TFLOPS fp16 on the GPU and integration with a 4-core CPU, at 45 Watts, with 6MB on-chip coherent cache, 128MB on-package last level cache, 16 GB DDR4, and 256GB SSD [1][2][3]. I would guess the total system power is under 150 Watts, as it is intended to compete with game consoles. This platform could in theory compare favorably with a system equipped with a Tesla M4 GPU [4], both in terms of inference throughput and total system power. Of course Skull Canyon was not designed for the data center, but neither was the GTX 980. We can look past this to the next generation of AMD Zen APUs which according to PS4K and Xbox rumors should raise the performance bar quite a bit higher.

[1] http://www.anandtech.com/show/10152/intels-skull-canyon-nuc-is-officia
[2] http://ark.intel.com/products/93341/Intel-Core-i7-6770HQ-Processor-6M-Cache-up-to-3_50-GHz
[3] http://wccftech.com/intel-iris-pro-graphics-gamers/
[4] http://images.nvidia.com/content/tesla/pdf/nvidia-tesla-m4-datasheet.pdf

**Comment by soumith** [comment]
After talking to Rajat from the TensorFlow team, we thought that an automated Jenkins server will go a long way in maintaining these benchmarks, as well as benchmarking on new hardware. Rajat and Martin Wicke are helping set this stuff up here :)
After the initial setup gets going on a digits box, we can figure out how to add donated machines with different configurations from the community as slaves to the jenkins server.

**Comment by rajatmonga** [comment]
@hughperkins : @martinwicke is working with @soumith to set something like this up through our Jenkins setup. The idea would be to run these tests automatically every X days or so.

@andravin @naibaf7  I like the idea of running on other relevant hardware e.g. Intel and AMD GPUs that make sense. Not all libraries may support every vendor, but will be useful to have baselines from even a few.

**Comment by andravin** [comment]
@soumith Are accuracy / convergence tests back on the table? Per layer numeric accuracy and iterations to / accuracy of convergence are both informative measures. We do not really understand in detail how the former affects the latter, so both would be interesting.

All floating point calculations are approximations. fp16 quantization is an explicit approximation. Fast algorithms have different accuracy than naive algorithms. Bugs are very unstable approximations. ;-)

I have considered submitting a closed source "fp0" kernel to convnet-benchmarks. Only after results are released will I reveal that it just fills the output buffer with zeros. Then maybe everybody would appreciate the importance of accuracy tests.

**Comment by soumith** [comment]
@andravin accuracy / convergence tests are off the table atleast for the first deepmark release. There's simply not enough bandwidth / man-power to pull it off.

> I have considered submitting a closed source "fp0" kernel to convnet-benchmarks. Only after results are released will I reveal that it just fills the output buffer with zeros. Then maybe everybody would appreciate the importance of accuracy tests.

Convnet-benchmarks and deepmark are done in good faith and really are a tool for the community built by the community. There are several things that can be done to show an adversarial setting, but that would be pointless. Each of these frameworks are also used widely in other settings, making them have an implicit accuracy test.

**Comment by hughperkins** [comment]
@andravin 

If DeepMark works like convnet-benchmark worked, and I have no reason to think that it wont, then all the scripts are publicly available, in a relatively standard-ish format.  There is no reason why you couldnt set up a parallel repository, which runs accuracy tests against each library, and reports the results.

**Comment by hughperkins** [comment]
(actually, in general, in my opinion, it would be better to factorize deepmark into per-domain repos, like images, nlp, video, etc, and get them working one by one.  Otherwise it risks becoming a bit of a Doom3 :-D  Just my opinion though.... ).

**Comment by andravin** [comment]
@soumith I have been a software engineer for too long to believe that good intentions are enough to ensure accuracy. ;-) I do believe that rewarding frameworks for speed without regard for accuracy creates a moral hazard.

In order to believe that the community is taking care of convergence testing, I would want to see links to some learning curves for recent Imagenet winners (not Alexnet, a recent very deep model from the last year) using fp16, FFT, and Winograd kernels.

**Comment by bhack** [comment]
I think that also a miracle for a standard definition of a network specification and its weights could really help the whole ecosystem. Tensorflow is in a position that could try to keep the responsibility of the leadership for a standard with versioning but probably everybody want to eat its own food.

**Comment by hughperkins** [comment]
> I think that also a miracle for a standard definition of a network specification and its weights could really help the whole ecosystem.

Just as for andravin's request, this could also be done by an individual contributor, orthogonally to all other efforts: simply pick one format, or make up a new one (obligatory xckd :-D https://m.xkcd.com/927/ ), and create converters to/from the most important frameworks. But ... isnt caffe zoo kind of already a de facto standard for this?

**Comment by bhack** [comment]
So protobuf is the de facto standard? I believe only in something more formal where all major framework are involved by design without to/from converters (or only optional for back compatibility). Cudnn API in some way is the only de facto standard.

**Comment by scott-gray** [comment]
For training with existing fp16 kernels you'll likely need a few tricks.  To allow weight updates to proceed there needs to be enough overlap in mantissa and the weight for the summation to have any effect.   Batch norm helps a lot here.  You can also use stochastic rounding, which allows weight updates even when there is no effective overlap.  But probably the best way is to leverage the fact that the filter tensors are relatively small and just keep them in fp32.  You can quantize just prior to the conv operation.  If you're already doing some kind of transform or dimshuffle you can just bake that quantization into that at no cost.

Batch norm also helps a lot with keeping the activations and delta's bounded within the range that fp16 can represent.  Weight initialization that factors in the fan-in also helps here.  Winograd transforms tend to widen the dynamic range of the activations/deltas so that's something to look out for too.

On upcoming pascal hardware we'll really need to fully explore the limits of fp16 accumulate.  I'm guessing some of the faster winograd transforms might break down here.  It's a shame that nvidia didn't make that fp16x2 instruction an inner product accumulate like the new int8x4 instruction will be.  With that you could multiply two sets of fp16 values and accumulate to a single fp32 register.

Anyway, I think with a bit of care you can generally achieve exactly the same accuracy with fp16 as you can with fp32.  I've even seen the reduced precision act as a regularizer to increase accuracy.

**Comment by hughperkins** [comment]
> So protobuf is the de facto standard? I believe only in something more formal where all major framework are involved by design without to/from converters

Well, I'm not sure about the 'without to/form converters' part, but torch [already supports protobuf](https://github.com/szagoruyko/loadcaffe).  As does caffe.  mxnet https://github.com/dmlc/mxnet/blob/master/tools/caffe_converter/README.md tensorflow https://github.com/ethereon/caffe-tensorflow

**Comment by bhack** [comment]
Yes but every framework write whatever they want with protobuf.
Converters don't avoid framework feature disaligment. Actually in production seems that cudnn versions defines what it is really important to introduce in API. It could be really better that with a common format this could be defined by design by frameworks i.e. [like in containers domain](https://www.opencontainers.org)

**Comment by hughperkins** [comment]
> Actually in production seems that cudnn versions defines what it is really important to introduce in API. It could be really better that with a common format this could be defined by design by frameworks i.e. like in containers domain

Without having looked at it (because I dont want to be exposed to the IP, if I can avoid it), I would think that the cudnn interfaces are very good.  However, note that to see their api, as far as I know, you have to click through a licensing agreement saying you understand it is their IP and so on, so I'm not sure that one could use the cudnn API as a standard?  (unless you could get agreement from NVIDIA of course).  By the way, you are hiding behind a pseudonym, I am curious what is your real name :-)

**Comment by bhack** [comment]
I will like a common roundtable like for containers format hosted by linux foundation.
I'm sure that this format need to be always a little bit behind state of the art but still very helpful in production. 
Probably @openai could has the "neutral status" to host a multi stakeholders design and maintenance.

P.s. I'm a simple GSoC 2016 mentor for OpenCV foundation :)

**Comment by hughperkins** [comment]
> Probably @openai could has the "neutral status" to host a multi stakeholders design and maintenance.

Interesting.  It does seem from http://openai.sourceforge.net/about.html that they are interested in specifying standards for ai-related tools

**Comment by hughperkins** [comment]
> Probably @openai could has the "neutral status" to host a multi stakeholders design and maintenance.

Interesting

**Comment by martinwicke** [comment]
I am doubtful that the sourceforge (sic, really?) thing is related to
openai (https://openai.com)

**Comment by hughperkins** [comment]
> I am doubtful that the sourceforge (sic, really?) thing is related to

yeah, hence why I deleted my original comment :-D

**Comment by martinwicke** [comment]
Sorry. Working off emails, didn't see. :)

**Comment by andravin** [comment]
@scott-gray I am sure you know the difference between an idea that seems like it should work and an experimental result that is publicly available.

> In order to believe that the community is taking care of convergence testing, I would want to see links to some learning curves for recent Imagenet winners (not Alexnet, a recent very deep model from the last year) using fp16, FFT, and Winograd kernels.

I am a _little_ bit surprised that nobody addressed this point.

**Comment by scott-gray** [comment]
@andravin I agree that we need this.   It's just hard to make it a priority over other things.  But I think with pascal coming out, there will be a real push to make fp16 work since the potential speedups will be so much more dramatic than they are now.

**Comment by forresti** [comment]
In terms of a standard format for representing DNN architectures and weights:

There is a tension between research and standardization. As mentioned above, there is some native support and/or conversion scripts for protobuf representations among caffe, torch, tensorflow, and mxnet. But, as research charges ahead, people in each framework/community are trying different types of new layers, solvers, weight initialization schemes, etc. Convolutions and SGD are very popular, but the finer points are still in flux. It may be too early to be worth trying to rigidly standardize all the frameworks.

**Comment by bhack** [comment]
@forresti Exactly what I have told. Standard need to stay behind state of the art. Is cudnn really cutting edge (excluding the performance point of view)?
No, arxiv is cutting edge, cudnn is behind and probably production. See also [Openvx at Kronos](https://www.khronos.org/openvx/) for CV.

**Comment by tqchen** [comment]
echo @bhack I have some thoughts in turns of standardize the frameworks. It is fundamentally hard for all frameworks to be forced into one. 

However, we might have enough idea on what are the common interface in a system that could be shared across frameworks, CuDNN is a perfect example of these. The possible things I can think of includes:

- Operator interface for registering kernels on different types of devices.
- Dependency scheduler to schedule things across devices.
- Data loading module to keep up with pre-processing pipeline.
- Representation of computation graph.
- Storage format of tensors.

I can see the finer we can break them, there is a better chance to see chance of sharing of some modules between frameworks, and move the field as a whole. I would love to see such discussion happen, and possibly in a neutral way

**Comment by hughperkins** [comment]
> However, we might have enough idea on what are the common interface in a system that could be shared across frameworks, CuDNN is a perfect example of these.

On a somewhat related note, I was thinking about how we could get portable dnn libraries, across multiple hardware, not just cuda hardware. We already have portable implementations of caffe, torch, maybe some others, by writing the kernels in OpenCL. However, the performance of the convolutions is arguably not as fast as one might achieve by using the hardware-level optimizations in winograd, and perhaps cudnn.

I'm thinking that it might be useful to have an api at exactly the cudnn level, so that libraries will automatically call cudnn on cuda platforms, or eg greentea on amd platforms, or some other convolutional library on eg Intel platforms. I made a pretty slide of this here :-)

https://raw.githubusercontent.com/hughperkins/project-ideas/master/img/openclconv1c.png

And note that this architecture would say nothing about who writes the hw-level implementation, eg one could imagine that the end-user could install winograd kernels on their machine, and caffe, mxnet, torch etc would automatically make use of them, without the dnn library authors even needing to know about this, or compile-time link with them:

https://raw.githubusercontent.com/hughperkins/project-ideas/master/img/openclconv2.png

**Comment by naibaf7** [comment]
@hughperkins 
Thanks for including my efforts :) not quite there yet with performance (writing the autotuning code at this instant).

I think this is a very nice idea. At least the convolutions should be somewhat standardized as they are the most performance critical part and about the same in every framework.

**Comment by lamblin** [comment]
> Well, I'm not sure about the 'without to/form converters' part, but torch [already supports protobuf](https://github.com/szagoruyko/loadcaffe).  As does caffe.  mxnet https://github.com/dmlc/mxnet/blob/master/tools/caffe_converter/README.md tensorflow https://github.com/ethereon/caffe-tensorflow
> 

[sklearn-theano](https://github.com/sklearn-theano/sklearn-theano) also has a [converter](https://github.com/sklearn-theano/sklearn-theano/blob/master/sklearn_theano/feature_extraction/caffe/caffemodel.py).

-- 
Pascal

**Comment by andravin** [comment]
While I find the discussion of standardization interesting and important, I really do not need any of that to do basic convergence testing.

For example, if I had Neon configuration files for Resnet or Inception that were known to give state of the art results when trained with fp32 direct convolution kernels, and access to a multi-GPU machine, then I could easily substitute the Winograd fp32 F(2x2,3x3) kernels for all the convolutions and compare the learning curve to the original. As simple as this experiment is, nobody has ever reported these results. Then you run the experiment again with F(4x4,3x3) kernels.

As @scott-gray mentioned, switching to fp16 weights might need some extra tricks. But all of this could be done within Neon.

Likewise the comparison of cuDNN direct convolution fp32, direct convolution fp16, and FFT should be possible entirely within TensorFlow.

Comparison of results between different frameworks might be a bit of an apples to oranges comparison, but within-framework comparison of different algorithms gives us a measure of the usefulness of state of the art kernels. Also it would force us to nail down the tricks that make them perform best.

I would think it is in the best interest of framework maintainers to facilitate these experiments.

**Comment by bhack** [comment]
This @vrv [comment] (https://github.com/tensorflow/tensorflow/pull/2002#issuecomment-218783825) it is quite auto explicative of some of the standardization posts in this thread

**Comment by vrv** [comment]
1) eh, software is tricky, nothing to do with standardization
2) this conversation has ratholed from its original intent to discuss benchmarks, though the discussion is interesting.

**Comment by bhack** [comment]
1) Yes but why all the frameworks need to do it on cudnn? :)
2) Sure but there is no space for this cause frameworks still don't has started a discussion this topic.
3) One hardware vendor it is doing a standard de facto at some level.

**Comment by andravin** [comment]
@vrv Convergence speed vs. accuracy tests _are_ benchmarks. But the idea has no traction here so I will do it somewhere else.

**Comment by soumith** [comment]
@andravin they ARE benchmarks, and actually my early versions of deepmark iterated on convergence speed. But this requires way more effort than is available, so i converged onto something a bit cheaper overall...

**Comment by vrv** [comment]
My comment on ratholing was about standardization of APIs and frameworks -- I do think convergence speed and accuracy tests are important and we should eventually have them.

**Comment by bhack** [comment]
Yes we have diverted the discussion at some point. I hope that the framework with more starts on github on day could restart that topic in a more appropriate context than this one.

**Comment by andravin** [comment]
@soumith yes convergence tests of course require a lot of compute time. can we at least incorporate numeric accuracy tests with synthetic data, as I think @vrv mentioned earlier?

**Comment by soumith** [comment]
@andravin no. not in V1.

**Comment by scott-gray** [comment]
One point about synthetic accuracy tests is that it doesn't necessarily correspond to final test accuracy.  As I was saying earlier, low precision can sometimes produce better results.  I'll quote the conclusion of this [paper](http://arxiv.org/pdf/1510.03009v3.pdf) here:

> Low precision prevents the optimizer from finding solutions that require a lot of precision, which correspond to very thin (high curvature) critical points, and these minima are more likely to correspond to overfitted solutions then broad minima (there are more functions that are compatible with such solutions, corresponding to a smaller description length and thus better generalization).

**Comment by vrv** [comment]
Step 1: define important models / benchmarks
Step 2: make sure people's definitions of models match so the numbers are actually comparable
Step 3: care and compare multi-step convergence

I don't think you can have 2 without 1, and 3 without 2.  So the order @soumith is pursuing this seems fine.

**Comment by andravin** [comment]
The approach I would take is:

A) kernel efficiency versus numeric accuracy testing
B) whole network convergence A/B testing (where you vary one of 1. kernel, 2. precision, 3. batch size, etc...). 
C) framework profiling (measure time spent in low level kernels and compare with total run time).

I think that will give us a more detailed picture of the factors that contribute to neural network performance.

**Comment by soumith** [comment]
So, getting back to the discussion. 
I have volunteers from a few frameworks, looking for some.

Volunteers available
================
Torch: Soumith, @SeanNaren , Shubo Sengupta (Baidu)
TensorFlow: @vrv and @martinwicke 
Neon: Evren Tumer @nervetumer
Chainer: @delta2323 and team

Volunteers needed
================
Caffe / NVIDIA-Caffe: [Need volunteers] cc: @bhack @Yangqing @lukeyeager 
MXNet: [Need volunteers] @antinucleon and @tqchen , is there anyone in the MXNet community who will do this?
Theano / Lasagne / Keras: [Need volunteers] cc: @nouiz @benanne @pranv 

If I dont find volunteers for some frameworks, I'll (without guarantee) try to do them myself.

**Comment by antinucleon** [comment]
@soumith OK. I will do it

**Comment by andravin** [comment]
The 3 different benchmarks I outlined A) kernel speed/accuracy, B) network convergence, C) framework profiling are independent of each other, can be done in any order, and can even be separate projects.

convnet-benchmarks and the proposed deepmark are a very coarse grained version of C) framework profiling (they just measure the whole network forward/backward iteration time). You could get more information out of the benchmarks if you drilled down to per layer timings, and even better, drill down to kernel timings. Then we would know why one framework is faster than another, which the current tests do not tell us.

A good graph computation framework would benchmark itself continuously under normal operation. I do not know the extent to which TensorFlow and Neon do this already, but it is not hard to instrument code so that all layers and kernel calls are timed and then generate summary statistics. Of course memory use could also be measured.

So I see a good framework profiling benchmark:

1) assisting framework maintainers to instrument their code for auto-benchmarking 
2) providing standard synthetic data generators for each framework to wrap as an input layer
3) creating per-framework configuration files for each test
4) running tests and collecting the benchmark data logged by the framework
5) generating reports from benchmark data

Instrumenting frameworks for auto-benchmarking would also help users tune networks for fast operation, and help framework maintainers identify what to optimize. As such the deepmark profiling report would be a tool used not just for benchmark contests, but part of the everyday machine learning workflow.

**Comment by hughperkins** [comment]
Andrew,

No-one here is seeing that we dont want correctness.  I think it's fair to say we would all be happy to see tests for correctness.  I think it's also fair to say that most frameworks already have unit tests and other tests for correctness.  I think it's fair to say that anything that hasnt been tested generally doesnt work, so I would be very surprised to see a framework without any kind of testing :-)

I think the only question is: to add correctness tests alongside the exact models in this benchmark, and to do it in a way so you can show numerical equivalence, down to say 4 significant figures, across all weights and activations, is a phenomenal amount of work, that goes far beyond what is already in place for our correctness checking.  If you are saying that you are going to handle this effort, and all we have to do is sit back and look at the results, then I dont think there is any discussion required :-)

**Comment by f0k** [comment]
You can count me in again for Theano (either vanilla or using Lasagne), but I'd prefer not to be the only one.

**Comment by pranv** [comment]
I can help @f0k with Theano, but only after 23rd.

**Comment by andravin** [comment]
@hughperkins please read my posts more carefully. I wrote that deepmark could be a framework profiling tool rather than just a forward/backward iteration timing tool.

TensorFlow is now instrumented for profiling: https://github.com/tensorflow/tensorflow/issues/1824

Numeric accuracy tests, on the other hand, can be a separate project and should be at the level of individual kernels.

**Comment by hughperkins** [comment]
> @hughperkins please read my posts more carefully.  I wrote that deepmark could be a framework profiling tool rather than just a forward/backward iteration timing tool.

Hi Andrew, fair point :-D

**Comment by jbalma** [comment]
I apologize if my comment further derails this thread but this seems like an appropriate time to bring it up.   I wonder if anyone is interested in strong scaling the time to convergence mentioned earlier? This would require what andravin calls tasks A and B above. It will soon be required to demonstrate convergence and the resulting accuracy at various network/dataset scales.

I'm less worried about functional profiling since external tools exist to do this fairly well and the algorithmic bottlenecks in sgd are a good first order place to optimize. It's also good to see mentioned the effectiveness of quantization (1bit and 8bit sgd) as I suspect these will become important features in frameworks to come especially with regard to scaling. 

I've been working with CNTK on many node systems recently and have become interested in how other platform's model parallelism scales in terms of time to solution (which is a tricky thing the community needs to define sooner or later). I really like the idea of having a common number to compare across frameworks (samples per second / epoch work well but don't capture wall time ) , which has inspired needed attention by developers on performance at the particular scale which is most used today. The main issue becomes larger and more complex models need to be used to demonstrate strong scaling as system size increases beyond a few gpus. I've run into this problem on large systems because there just isn't enough work to distribute using current problem sizes. 

That said, it would be incredibly useful to have a set of even two or three (small medium large) benchmarks (convolution + fully connected) which use the same training data but at distinct scales in terms of number of features, input dimension and hiddens such that we could measure and estimate how well the model/data parallelism preforms across MANY cpus and gpus and it would help us pin down performance bottlenecks as a side effect. The convergence and accuracy criteria then become important though, as well as going beyond the current memory requirements. 

This is a dream list of criteria to meet and no small task but I wonder if anyone else would be interested in such an effort or perhaps it's already going on somewhere? 

I think the profiling and benchmark efforts here are great and incredibly beneficial in creating standards / best practices across frameworks going forward. Thanks for all the work and for a great thread to read at the least!

**Comment by soumith** [comment]
Excellent @pranv and @f0k . @craffel was interested in partly doing Theano benchmarking as well.

**Comment by gnawice** [comment]
I'd hope this also supports CPU training/testing. 
My mojo-cnn project seems to be relatively fast on CPU for smaller problems and i know intel is still trying to push CPU use with their DNN wrappers for MKL to better compete with GPUs.

**Comment by andravin** [comment]
@jbalma yes I find strong scaling the training problem to be very interesting, and of course that is an active area of research. I hope you find a good forum for pursuing it. Let me know if I can help.

I would also point out with regards to profiling, an external profiling tool will of course not be able to correlate timings with the graph structure of the network, so it cannot group statistics by layer, which is essential for understanding the performance of the network. I think all successful machine learning frameworks will be instrumented for profiling eventually, and tensorflow appears to be taking the lead there. Now imagine the power of being able to compare network performance graphs across different frameworks, because they all were designed to output profiler stats in the same format.

**Comment by hughperkins** [comment]
> I think all successful machine learning frameworks will be instrumented for profiling eventually, and tensorflow appears to be taking the lead there

*cough* [cltorch profiling and instrumentation](https://github.com/hughperkins/cltorch#profiling-tools) *cough* :-P

**Comment by f0k** [comment]
>> I think all successful machine learning frameworks will be instrumented for profiling eventually, and tensorflow appears to be taking the lead there

> [cltorch profiling and instrumentation](https://github.com/hughperkins/cltorch#profiling-tools) :-)

`THEANO_FLAGS=profile=1` works since [at least 2013](https://github.com/Theano/Theano/commits/master/doc/tutorial/profiling.txt). Caveat: It profiles both CPU and GPU operations, so it times things "from outside", so it needs to be combined with `CUDA_LAUNCH_BLOCKING=1` to make GPU operations synchronous, so it cannot profile a real-life program. We could [time kernels with CUDA events](https://devblogs.nvidia.com/parallelforall/how-implement-performance-metrics-cuda-cc/) instead, but that would limit timings to things happening in a CUDA stream.

> I would also point out with regards to profiling, an external profiling tool will of course not be able to correlate timings with the graph structure of the network, so it cannot group statistics by layer, which is essential for understanding the performance of the network.

If we want metrics per layer (not per operation, or per operation type), we'll need to figure out how/where to introduce timing checkpoints in the graph without hindering optimization. That's a fundamental problem for frameworks based on computation graphs: They are free to rearrange operations across layer boundaries, so you cannot necessarily correlate the optimized graph with the network layers at all.

**Comment by nouiz** [comment]
On Thu, May 19, 2016 at 7:46 AM, Jan Schlüter <notifications@github.com>
wrote:

> I think all successful machine learning frameworks will be instrumented
> for profiling eventually, and tensorflow appears to be taking the lead there
>
> cltorch profiling and instrumentation
> <https://github.com/hughperkins/cltorch#profiling-tools> :-)
>
> THEANO_FLAGS=profile=1 works since at least 2013
> <https://github.com/Theano/Theano/commits/master/doc/tutorial/profiling.txt>.
> Caveat: It profiles both CPU and GPU operations, so it times things "from
> outside", so it needs to be combined with CUDA_LAUNCH_BLOCKING=1 to make
> GPU operations synchronous, so it cannot profile a real-life program. We
> could time kernels with CUDA events
> <https://devblogs.nvidia.com/parallelforall/how-implement-performance-metrics-cuda-cc/>
> instead, but that would limit timings to things happening in a CUDA stream.
>
As the CPU node are already blocking, we could just add this as a second
timing columns that would be non-0 only for the new gpu-backend.


> I would also point out with regards to profiling, an external profiling
> tool will of course not be able to correlate timings with the graph
> structure of the network, so it cannot group statistics by layer, which is
> essential for understanding the performance of the network.
>
> If we want metrics per layer (not per operation, or per operation type),
> we'll need to figure out how/where to introduce timing checkpoints in the
> graph without hindering optimization. That's a fundamental problem for
> frameworks based on computation graphs: They are free to rearrange
> operations across layer boundaries, so you cannot necessarily correlate the
> optimized graph with the network layers at all.
>
We could reject optimization that merge node from 2 layers to have exact
per layer timing. Otherwise we could just let layer to be merged in the
profiler.

As always, we can always do better!

**Comment by hughperkins** [comment]
or, we could use the existing per-layer timings method, which seems to me to be a reasonably rigorous way of doing it, and avoids fudge-factors based on differing opinions on how to do in-situ timings.  like:
- do we call synchronize() between layers?
- wallclock timings?  kernel profiling timings?
I'm sure there are a bunch more of such questions....

**Comment by f0k** [comment]
> We could reject optimization that merge node from 2 layers to have exact per layer timing.

The event-based timing could be realized as Ops that insert events into the CUDA stream. If inserted between layers, these Ops would naturally block any optimizations across layers if existing optimizers are not adapted to ignore them. But again, this would prevent real in-situ timings, because the graph might not be fully optimized.

> I'm sure there are a bunch more of such questions....

Yes... in-situ timings would be the best, but depending on the framework, we cannot get per-layer timings without changing the process, and maybe not even distinguish the forward and backward pass, just training (fw+bw) and inference (fw only). For the start, we should probably have per-epoch or per-batch timings only.

**Comment by soumith** [comment]
On the Caffe side, Mike Houston and team at NVIDIA have agreed to do the benchmark scripts. So that concludes all the volunteers for each framework :)

**Comment by naibaf7** [comment]
@soumith 
I might do some adaptions to the Caffe scripts for OpenCL/Greentea-libDNN if required :)

**Comment by soumith** [comment]
Sure :)

**Comment by naibaf7** [comment]
@soumith 
I know we're on the edge of switching to a new benchmark system, but it would be great if you could give this a shot: https://github.com/soumith/convnet-benchmarks/pull/106
Thanks :)

**Comment by hughperkins** [comment]
Opinion: the new system should be a collection of specialized repos, each with their own repo, being:

- images: owner Soumith, basically this one
- video: owner: ???
- nlp: owner: ???
... etc.  Therefore, I see no reason for this repo disappearing in any way, shape or form, personally :-)

**Comment by hughperkins** [comment]
forresti wrote:

> Agreed. Soumith's current benchmarks are useful, but they mainly evaluate "who can make the thinnest wrapper around cuDNN, Neon, or similar?"

On a somewhat related note, since GEMM seems to be at the heart of convolution.  It's used by fft, winograd, im2col, and presumably also implicit gemm.  So, could it be worth having some GEMM benchmarks?  For OpenCL, there are at least 3 GEMM implementations I know of, ie: clBLAS, clBLAST, ViennaCL.

**Comment by bhack** [comment]
@hughperkins Do you mean something like https://github.com/dividiti/gemmbench?

**Comment by hughperkins** [comment]
@bhack 

Possibly.  I'm not sure what I mean to be honest.  I'm not sure that's quite exactly what I was thinking of.  I was thinking of something more like the simple tables in convnet-benchmarks, but comparing these 5 or so GEMM implementations.  Presumably, the actual workloads, if we are targeting convolution, should be workloads sampled by running a forwards-backwards batch through a few common convolutional models, such as those currently in convnet-benchmarks.

**Comment by naibaf7** [comment]
@hughperkins 
I actually think GEMM will be a bit less at the heart of convolutions as we move forward. At least not pure GEMM as implemented in BLAS libraries.
There are many nice optimization and memory coalescing possibilities when implementing kernel fusion, that the GEMM only remains at the core (register blocked GEMM, at most shared memory GEMM).
I am just right now doing these optimizations on libDNN to get closer to cuDNN scores.

Also, what makes benchmarking the GEMMs more difficult is that their performance differs a lot on a variety of devices. And then the GEMMs can also be autotuned, which works more or less depending on what BLAS and architecture combination is used.

I think benchmarking on the network and layer level is enough for the DeepMark project.

**Comment by hughperkins** [comment]
> There are many nice optimization and memory coalescing possibilities when implementing kernel fusion, that the GEMM only remains at the core (register blocked GEMM, at most shared memory GEMM).
I am just right now doing these optimizations on libDNN to get closer to cuDNN scores.

Ah, sounds good :-)

**Comment by hughperkins** [comment]
Hi Andrew, So ... guess what?   I wrote a correctness checking script :-)  And ... ironically enough... it targets Neon :-D  Because I need it for testing the OpenCL port.

It outputs for each layer:
- average forward time
- average backwards time
- average delta between each output value and a cpu-calculated reference value, assumed correct, since python
uses float64 (I think?)
- ditto for weights gradient
- ditto for input gradient

Example [results](https://github.com/hughperkins/neon-benchmarks/tree/master/results), for neon on Titan X, using [vgga](https://github.com/hughperkins/neon-benchmarks/blob/master/models/vgga.py) model:

[Maxwell kernels, CUDA](https://github.com/hughperkins/neon-benchmarks/blob/master/results/neonbase_maxwell_vgga.txt):
```
Layer 0: SKIPPING
Layer 1: fprop=0.012 bprop=0.032 eps_O=1e-05 eps_gradW=1e-02 eps_gradI=3e-05
Layer 2: fprop=0.009 bprop=0.021 eps_O=5e-05 eps_gradW=2e-03 eps_gradI=6e-05
Layer 3: fprop=0.017 bprop=0.036 eps_O=1e-04 eps_gradW=3e-03 eps_gradI=8e-05
Layer 4: fprop=0.007 bprop=0.016 eps_O=8e-05 eps_gradW=6e-04 eps_gradI=2e-04
Layer 5: fprop=0.015 bprop=0.031 eps_O=9e-05 eps_gradW=8e-04 eps_gradI=1e-04
Layer 6: fprop=0.005 bprop=0.010 eps_O=8e-05 eps_gradW=4e-04 eps_gradI=9e-05
Layer 7: fprop=0.005 bprop=0.010 eps_O=8e-05 eps_gradW=4e-04 eps_gradI=9e-05
```
[Kepler kernels, CUDA](https://github.com/hughperkins/neon-benchmarks/blob/master/results/neonbase_kepler_vgga.txt):
```
Layer 0: SKIPPING
Layer 1: fprop=0.032 bprop=0.158 eps_O=9e-06 eps_gradW=1e-03 eps_gradI=2e-05
Layer 2: fprop=0.033 bprop=0.110 eps_O=2e-05 eps_gradW=3e-04 eps_gradI=2e-05
Layer 3: fprop=0.067 bprop=0.222 eps_O=2e-05 eps_gradW=3e-04 eps_gradI=3e-05
Layer 4: fprop=0.033 bprop=0.111 eps_O=3e-05 eps_gradW=1e-04 eps_gradI=9e-05
Layer 5: fprop=0.066 bprop=0.222 eps_O=4e-05 eps_gradW=9e-05 eps_gradI=5e-05
Layer 6: fprop=0.016 bprop=0.053 eps_O=4e-05 eps_gradW=2e-05 eps_gradI=3e-05
Layer 7: fprop=0.016 bprop=0.053 eps_O=4e-05 eps_gradW=2e-05 eps_gradI=3e-05
```
[Kepler kernels, OpenCL](https://github.com/hughperkins/neon-benchmarks/blob/master/results/winogradcl_vgga.txt):
```
Layer 0: SKIPPING
Layer 1: fprop=0.039 bprop=0.173 eps_O=1e-05 eps_gradW=8e-04 eps_gradI=1e-05
Layer 2: fprop=0.039 bprop=0.124 eps_O=1e-05 eps_gradW=4e-04 eps_gradI=2e-05
Layer 3: fprop=0.073 bprop=0.237 eps_O=3e-05 eps_gradW=2e-04 eps_gradI=4e-05
Layer 4: fprop=0.038 bprop=0.125 eps_O=2e-05 eps_gradW=5e-05 eps_gradI=2e-05
Layer 5: fprop=0.073 bprop=0.238 eps_O=3e-05 eps_gradW=1e-04 eps_gradI=4e-05
Layer 6: fprop=0.022 bprop=0.069 eps_O=5e-05 eps_gradW=2e-05 eps_gradI=7e-05
Layer 7: fprop=0.021 bprop=0.067 eps_O=5e-05 eps_gradW=2e-05 eps_gradI=7e-05
```

https://github.com/hughperkins/neon-benchmarks

(Note: layer 0 should probably not be skipped for the CUDA kernels. I need to make the benchmark scripts not automatically skip anything with C < 4)

**Comment by hughperkins** [comment]
For the benchmarks with correctness checker, added stride and padding, so it can handle eg alexnet now:

https://github.com/hughperkins/neon-benchmarks/blob/master/results/alexnet_summary.md

## Nervana Neon CUDA/SASS Winograd kernels for Maxwell

[neon_maxwell](backends/neon_maxwell.py)
```
Layer 0: fprop=0.003 bprop=0.012 eps_O=4e-06 eps_gradW=5e-04 eps_gradI=2e-06
Layer 1: fprop=0.009 bprop=0.020 eps_O=9e-06 eps_gradW=4e-04 eps_gradI=3e-05
Layer 2: fprop=0.003 bprop=0.006 eps_O=5e-05 eps_gradW=3e-04 eps_gradI=8e-05
Layer 3: fprop=0.004 bprop=0.007 eps_O=5e-05 eps_gradW=7e-04 eps_gradI=1e-04
Layer 4: fprop=0.003 bprop=0.005 eps_O=1e-04 eps_gradW=3e-04 eps_gradI=6e-05
```

## Nervana Neon Kepler direct kernels, in CUDA

[neon_kepler](backends/neon_kepler.py)
```
Layer 0: SKIPPED
Layer 1: fprop=0.015 bprop=0.046 eps_O=9e-06 eps_gradW=2e-04 eps_gradI=3e-05
Layer 2: fprop=0.008 bprop=0.023 eps_O=2e-05 eps_gradW=3e-05 eps_gradI=5e-05
Layer 3: fprop=0.010 bprop=0.030 eps_O=4e-05 eps_gradW=3e-05 eps_gradI=2e-05
Layer 4: fprop=0.007 bprop=0.020 eps_O=3e-05 eps_gradW=2e-05 eps_gradI=2e-05
```

## OpenCL port of Nervana Neon Kepler direct kernels

[neoncl_direct](backends/neoncl_direct.py)
```
Layer 0: SKIPPED
Layer 1: fprop=0.016 bprop=0.049 eps_O=1e-05 eps_gradW=1e-04 eps_gradI=6e-05
Layer 2: fprop=0.008 bprop=0.024 eps_O=2e-05 eps_gradW=3e-05 eps_gradI=3e-05
Layer 3: fprop=0.010 bprop=0.031 eps_O=4e-05 eps_gradW=3e-05 eps_gradI=3e-05
Layer 4: fprop=0.007 bprop=0.021 eps_O=2e-05 eps_gradW=2e-05 eps_gradI=2e-05
```

However, there's still a couple of things we'd want, if we wanted to generalize this to other networks:
- some way of running the other frameworks from python, and / or
- some way of transmitting the tensors to the other frameworks

Actually, torch can [run from python](https://github.com/hughperkins/pytorch), as can theano, tensorflow, mxnet (I think?), caffe, DeepCL, chainer.  So maybe python is all that is needed???

**Comment by hughperkins** [comment]
Andrew, hmmm, just noticed, this is figure 4 in your paper.  Hadnt noticed that before :-D  Well, I hadnt read the paper earlier...   I see that you are using element max though.  I think this might be sensitive to outliers, and also larger for layers with more output values?  Maybe mean or average, possibly also with standard deviation, is good?

**Comment by f0k** [comment]
Possibly relevant: [Fathom: Reference Workloads for Modern Deep Learning Methods](http://arxiv.org/abs/1608.06581)

> Consequently, we have assembled Fathom: a collection of eight archetypal deep learning workloads for study. Each of these models comes from a seminal work in the deep learning community, ranging from the familiar deep convolutional neural network of Krizhevsky et al., to the more exotic memory networks from Facebook's AI research group. Fathom has been released online, and this paper focuses on understanding the fundamental performance characteristics of each model. We use a set of application-level modeling tools built around the TensorFlow deep learning framework in order to analyze the behavior of the Fathom workloads. We present a breakdown of where time is spent, the similarities between the performance profiles of our models, an analysis of behavior in inference and training, and the effects of parallelism on scaling.

**Comment by bhack** [comment]
See also Baidu Research [DeepBench](https://svail.github.io/DeepBench/)

**Comment by pedropgusmao** [comment]
Hi everyone, 
we will soon be benchmarking a NVIDIA DGX-1 and we are currently looking for ways to do it effectively. So if you have ever wondered how well the DGX-1 would perform on a specific task, let me know which. We will release all results online. 

For now, we have run some classic benchmarks with a K40/Maxwell Titan X/GTX 1080/ Pascal Titan X, using various architectures and frameworks. Results are available [here](http://add-for.com/blog/nvidia-dgx-1-supercomputer-join-our-community-based-deep-learning-benchmark/).
Again,  all suggestions on how to benchmark the DGX-1 are welcome. Thanks

**Comment by Yangqing** [comment]
@pedropgusmao from the curve it seems that you basically need to increase the workspace limit in Caffe:

https://github.com/BVLC/caffe/blob/80f44100e19fd371ff55beb3ec2ad5919fb6ac43/src/caffe/layers/cudnn_conv_layer.cpp#L113

The value was in order to support all platforms (old and new) but it usually comes slow for most recent cudnn with most recent hardware.

**Comment by pedropgusmao** [comment]
@Yangqing , thank you very much. Do you have any suggested values for workspace_limit_bytes considering both those GPUs and the DGX-1?  Again, thanks.

**Comment by vrv** [comment]
@pedropgusmao take a look at https://www.tensorflow.org/performance/performance_guide with TF 1.0.0.  I believe we are also working on benchmarks for some of these models, so you'll have comparison code at some point soon.

**Comment by pedropgusmao** [comment]
@vrv, thanks a lot! We will modify our code to follow those suggestions. We look forward to see your results.

**Comment by Julia991** [comment]
Guys, what do you think about this research? - [https://www.cleveroad.com/blog/github-vs-bitbucket-vs-gitlab-developers-research](https://www.cleveroad.com/blog/github-vs-bitbucket-vs-gitlab-developers-research)
What should I change?
